File: rust/src/config/policy.rs
use std::collections::HashSet;
use std::fs;
use std::io::{self, BufRead};
use std::sync::{Arc, RwLock};

use super::{InboundPolicy, InboundRule, ServiceConfig};
use super::labels::{LabelStore, match_labels};

/// Result of checking inbound access for a peer.
#[derive(Debug, Clone)]
pub struct PolicyResult {
    pub action: String,
    pub services: Vec<ServiceConfig>,
    pub rule_name: String,
    pub needs_zgrlan_verify: bool,
    pub zgrlan_peer: String,
}

struct CompiledEntry {
    rule: InboundRule,
    whitelist: HashSet<String>,
    list_path: String,
}

/// Evaluates inbound policy rules against peer public keys and labels.
/// Thread-safe for concurrent reads.
pub struct PolicyEngine {
    default_action: String,
    entries: RwLock<Vec<CompiledEntry>>,
    label_store: RwLock<Option<Arc<LabelStore>>>,
}

impl PolicyEngine {
    /// Create a PolicyEngine from an InboundPolicy.
    /// Whitelist files are loaded eagerly. LabelStore is optional.
    pub fn new(policy: &InboundPolicy) -> io::Result<Self> {
        let entries = Self::compile(&policy.rules)?;
        let default_action = if policy.default.is_empty() {
            "deny".to_string()
        } else {
            policy.default.clone()
        };
        Ok(Self {
            default_action,
            entries: RwLock::new(entries),
            label_store: RwLock::new(None),
        })
    }

    /// Create a PolicyEngine with a LabelStore for label-based matching.
    pub fn with_label_store(policy: &InboundPolicy, store: Arc<LabelStore>) -> io::Result<Self> {
        let pe = Self::new(policy)?;
        *pe.label_store.write().unwrap() = Some(store);
        Ok(pe)
    }

    /// Set or replace the label store.
    pub fn set_label_store(&self, store: Arc<LabelStore>) {
        *self.label_store.write().unwrap() = Some(store);
    }

    fn compile(rules: &[InboundRule]) -> io::Result<Vec<CompiledEntry>> {
        let mut entries = Vec::with_capacity(rules.len());
        for rule in rules {
            let mut entry = CompiledEntry {
                rule: rule.clone(),
                whitelist: HashSet::new(),
                list_path: String::new(),
            };
            if rule.match_config.pubkey.match_type == "whitelist" {
                let path = &rule.match_config.pubkey.path;
                entry.whitelist = load_pubkey_list(path)?;
                entry.list_path = path.clone();
            }
            entries.push(entry);
        }
        Ok(entries)
    }

    /// Evaluate the inbound policy for the given peer public key (raw 32 bytes).
    pub fn check(&self, pubkey: &[u8; 32]) -> PolicyResult {
        let pubkey_hex = hex::encode(pubkey);

        // Look up labels once
        let peer_labels: Vec<String> = {
            let store_guard = self.label_store.read().unwrap();
            match store_guard.as_ref() {
                Some(store) => store.labels(&pubkey_hex),
                None => vec![],
            }
        };

        let entries = self.entries.read().unwrap();
        for entry in entries.iter() {
            let rule = &entry.rule;

            let pubkey_matched;
            let mut needs_zgrlan = false;
            let mut zgrlan_peer = String::new();

            match rule.match_config.pubkey.match_type.as_str() {
                "any" => {
                    pubkey_matched = true;
                }
                "whitelist" => {
                    pubkey_matched = entry.whitelist.contains(&pubkey_hex);
                }
                "zgrlan" => {
                    pubkey_matched = true;
                    needs_zgrlan = true;
                    zgrlan_peer = rule.match_config.pubkey.peer.clone();
                }
                _ => continue,
            }

            if !pubkey_matched {
                continue;
            }

            // If the rule has label patterns, the peer must also match at least one
            if !rule.match_config.labels.is_empty()
                && !match_labels(&peer_labels, &rule.match_config.labels)
            {
                continue;
            }

            return PolicyResult {
                action: rule.action.clone(),
                services: rule.services.clone(),
                rule_name: rule.name.clone(),
                needs_zgrlan_verify: needs_zgrlan,
                zgrlan_peer,
            };
        }

        PolicyResult {
            action: self.default_action.clone(),
            services: vec![],
            rule_name: "default".into(),
            needs_zgrlan_verify: false,
            zgrlan_peer: String::new(),
        }
    }

    /// Reload whitelist files from disk.
    pub fn reload(&self) -> io::Result<()> {
        let mut entries = self.entries.write().unwrap();
        for entry in entries.iter_mut() {
            if !entry.list_path.is_empty() {
                entry.whitelist = load_pubkey_list(&entry.list_path)?;
            }
        }
        Ok(())
    }
}

fn load_pubkey_list(path: &str) -> io::Result<HashSet<String>> {
    let data = fs::read(path)?;
    let mut keys = HashSet::new();
    for (line_num, line) in data.lines().enumerate() {
        let line = line?;
        let trimmed = line.trim();
        if trimmed.is_empty() || trimmed.starts_with('#') {
            continue;
        }
        let lower = trimmed.to_lowercase();
        if lower.len() != 64 {
            return Err(io::Error::new(
                io::ErrorKind::InvalidData,
                format!("line {}: expected 64 hex chars, got {}", line_num + 1, lower.len()),
            ));
        }
        if hex::decode(&lower).is_err() {
            return Err(io::Error::new(
                io::ErrorKind::InvalidData,
                format!("line {}: invalid hex", line_num + 1),
            ));
        }
        keys.insert(lower);
    }
    Ok(keys)
}

#[cfg(test)]
mod tests {
    use super::*;
    use super::super::{MatchConfig, PubkeyMatch};
    use std::sync::Arc;

    fn hex_key(hex_str: &str) -> [u8; 32] {
        let mut key = [0u8; 32];
        let bytes = hex::decode(hex_str).unwrap();
        key[..bytes.len()].copy_from_slice(&bytes);
        key
    }

    fn make_policy(default: &str, rules: Vec<InboundRule>) -> InboundPolicy {
        InboundPolicy {
            default: default.into(),
            revalidate_interval: String::new(),
            rules,
        }
    }

    fn any_rule(name: &str, action: &str) -> InboundRule {
        InboundRule {
            name: name.into(),
            match_config: MatchConfig { pubkey: PubkeyMatch {
                match_type: "any".into(), path: String::new(), peer: String::new(),
            }, labels: vec![] },
            services: vec![ServiceConfig { proto: "*".into(), port: "*".into() }],
            action: action.into(),
        }
    }

    #[test]
    fn test_any_match() {
        let policy = make_policy("deny", vec![any_rule("allow-all", "allow")]);
        let pe = PolicyEngine::new(&policy).unwrap();

        let result = pe.check(&hex_key("abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789"));
        assert_eq!(result.action, "allow");
        assert_eq!(result.rule_name, "allow-all");
    }

    #[test]
    fn test_whitelist() {
        let dir = std::env::temp_dir().join("zgrnet_policy_test");
        let _ = std::fs::create_dir_all(&dir);
        let list_path = dir.join("trusted.txt");
        let trusted_key = "abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789";
        std::fs::write(&list_path, format!("# Trusted\n{trusted_key}\n")).unwrap();

        let policy = make_policy("deny", vec![InboundRule {
            name: "trusted".into(),
            match_config: MatchConfig { pubkey: PubkeyMatch {
                match_type: "whitelist".into(),
                path: list_path.to_str().unwrap().into(),
                peer: String::new(),
            }, labels: vec![] },
            services: vec![ServiceConfig { proto: "tcp".into(), port: "80,443".into() }],
            action: "allow".into(),
        }]);
        let pe = PolicyEngine::new(&policy).unwrap();

        let result = pe.check(&hex_key(trusted_key));
        assert_eq!(result.action, "allow");

        let result = pe.check(&hex_key("0000000000000000000000000000000000000000000000000000000000000001"));
        assert_eq!(result.action, "deny");
        assert_eq!(result.rule_name, "default");

        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn test_zgrlan() {
        let policy = make_policy("deny", vec![InboundRule {
            name: "company".into(),
            match_config: MatchConfig { pubkey: PubkeyMatch {
                match_type: "zgrlan".into(),
                path: String::new(),
                peer: "company.zigor.net".into(),
            }, labels: vec![] },
            services: vec![ServiceConfig { proto: "tcp".into(), port: "80,443".into() }],
            action: "allow".into(),
        }]);
        let pe = PolicyEngine::new(&policy).unwrap();

        let result = pe.check(&hex_key("abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789"));
        assert_eq!(result.action, "allow");
        assert!(result.needs_zgrlan_verify);
        assert_eq!(result.zgrlan_peer, "company.zigor.net");
    }

    #[test]
    fn test_default_allow() {
        let policy = make_policy("allow", vec![]);
        let pe = PolicyEngine::new(&policy).unwrap();
        let result = pe.check(&[0u8; 32]);
        assert_eq!(result.action, "allow");
    }

    #[test]
    fn test_default_deny() {
        let policy = make_policy("deny", vec![]);
        let pe = PolicyEngine::new(&policy).unwrap();
        let result = pe.check(&[0u8; 32]);
        assert_eq!(result.action, "deny");
    }

    #[test]
    fn test_priority() {
        let dir = std::env::temp_dir().join("zgrnet_policy_prio_test");
        let _ = std::fs::create_dir_all(&dir);
        let list_path = dir.join("trusted.txt");
        let key = "abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789";
        std::fs::write(&list_path, format!("{key}\n")).unwrap();

        let policy = make_policy("deny", vec![
            InboundRule {
                name: "trusted-full".into(),
                match_config: MatchConfig { pubkey: PubkeyMatch {
                    match_type: "whitelist".into(),
                    path: list_path.to_str().unwrap().into(),
                    peer: String::new(),
                }, labels: vec![] },
                services: vec![ServiceConfig { proto: "*".into(), port: "*".into() }],
                action: "allow".into(),
            },
            any_rule("any-limited", "allow"),
        ]);
        let pe = PolicyEngine::new(&policy).unwrap();

        let result = pe.check(&hex_key(key));
        assert_eq!(result.rule_name, "trusted-full");

        let result = pe.check(&hex_key("0000000000000000000000000000000000000000000000000000000000000001"));
        assert_eq!(result.rule_name, "any-limited");

        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn test_label_match() {
        let store = Arc::new(LabelStore::new());
        let trusted_key = "abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789";
        store.set_labels(trusted_key, &["host.zigor.net/trusted".into()]);

        let policy = make_policy("deny", vec![InboundRule {
            name: "label-trusted".into(),
            match_config: MatchConfig {
                pubkey: PubkeyMatch {
                    match_type: "any".into(), path: String::new(), peer: String::new(),
                },
                labels: vec!["host.zigor.net/trusted".into()],
            },
            services: vec![ServiceConfig { proto: "tcp".into(), port: "80,443".into() }],
            action: "allow".into(),
        }]);
        let pe = PolicyEngine::with_label_store(&policy, store).unwrap();

        let result = pe.check(&hex_key(trusted_key));
        assert_eq!(result.action, "allow");
        assert_eq!(result.rule_name, "label-trusted");

        // Unknown key has no labels -> deny
        let result = pe.check(&hex_key("0000000000000000000000000000000000000000000000000000000000000001"));
        assert_eq!(result.action, "deny");
    }

    #[test]
    fn test_label_wildcard() {
        let store = Arc::new(LabelStore::new());
        let key = "abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789";
        store.set_labels(key, &["company.zigor.net/employee".into()]);

        let policy = make_policy("deny", vec![InboundRule {
            name: "company-all".into(),
            match_config: MatchConfig {
                pubkey: PubkeyMatch {
                    match_type: "any".into(), path: String::new(), peer: String::new(),
                },
                labels: vec!["company.zigor.net/*".into()],
            },
            services: vec![ServiceConfig { proto: "*".into(), port: "*".into() }],
            action: "allow".into(),
        }]);
        let pe = PolicyEngine::with_label_store(&policy, store).unwrap();

        let result = pe.check(&hex_key(key));
        assert_eq!(result.action, "allow");
    }

    #[test]
    fn test_no_label_store() {
        // Without a LabelStore, label rules skip
        let policy = make_policy("deny", vec![InboundRule {
            name: "label-only".into(),
            match_config: MatchConfig {
                pubkey: PubkeyMatch {
                    match_type: "any".into(), path: String::new(), peer: String::new(),
                },
                labels: vec!["host.zigor.net/trusted".into()],
            },
            services: vec![ServiceConfig { proto: "*".into(), port: "*".into() }],
            action: "allow".into(),
        }]);
        let pe = PolicyEngine::new(&policy).unwrap();

        let result = pe.check(&hex_key("abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789"));
        assert_eq!(result.action, "deny");
    }

    #[test]
    fn test_reload() {
        let dir = std::env::temp_dir().join("zgrnet_policy_reload_test");
        let _ = std::fs::create_dir_all(&dir);
        let list_path = dir.join("trusted.txt");
        let key1 = "abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789";
        let key2 = "0000000000000000000000000000000000000000000000000000000000000001";
        std::fs::write(&list_path, format!("{key1}\n")).unwrap();

        let policy = make_policy("deny", vec![InboundRule {
            name: "trusted".into(),
            match_config: MatchConfig { pubkey: PubkeyMatch {
                match_type: "whitelist".into(),
                path: list_path.to_str().unwrap().into(),
                peer: String::new(),
            }, labels: vec![] },
            services: vec![ServiceConfig { proto: "*".into(), port: "*".into() }],
            action: "allow".into(),
        }]);
        let pe = PolicyEngine::new(&policy).unwrap();

        assert_eq!(pe.check(&hex_key(key2)).action, "deny");

        std::fs::write(&list_path, format!("{key1}\n{key2}\n")).unwrap();
        pe.reload().unwrap();

        assert_eq!(pe.check(&hex_key(key2)).action, "allow");

        let _ = std::fs::remove_dir_all(&dir);
    }
}

File: zig/src/config/policy.zig
//! Inbound policy engine: evaluates rules against peer public keys and labels.

const std = @import("std");
const mem = std.mem;
const Allocator = std.mem.Allocator;
const types = @import("types.zig");
const labels_mod = @import("labels.zig");
const InboundPolicy = types.InboundPolicy;
const InboundRule = types.InboundRule;
const ServiceConfig = types.ServiceConfig;
const LabelStore = labels_mod.LabelStore;
const matchLabels = labels_mod.matchLabels;

pub const PolicyResult = struct {
    action: []const u8,
    services: []const ServiceConfig,
    rule_name: []const u8,
    needs_zgrlan_verify: bool,
    zgrlan_peer: []const u8,
};

const CompiledEntry = struct {
    rule: *const InboundRule,
    whitelist: ?std.StringHashMapUnmanaged(void),
    list_path: []const u8,
};

/// Evaluates inbound policy rules against peer public keys and labels.
pub const PolicyEngine = struct {
    default_action: []const u8,
    entries: std.ArrayListUnmanaged(CompiledEntry) = .{},
    allocator: Allocator,
    label_store: ?*const LabelStore = null,

    pub fn init(allocator: Allocator, policy: *const InboundPolicy) !PolicyEngine {
        var pe = PolicyEngine{
            .default_action = if (policy.default.len == 0) "deny" else policy.default,
            .allocator = allocator,
        };
        errdefer pe.deinit();

        for (policy.rules) |*rule| {
            var entry = CompiledEntry{
                .rule = rule,
                .whitelist = null,
                .list_path = "",
            };

            if (mem.eql(u8, rule.@"match".pubkey.@"type", "whitelist")) {
                entry.list_path = rule.@"match".pubkey.path;
                entry.whitelist = try loadPubkeyList(allocator, rule.@"match".pubkey.path);
            }

            try pe.entries.append(allocator, entry);
        }

        return pe;
    }

    /// Create a PolicyEngine with a LabelStore for label-based matching.
    pub fn initWithLabelStore(allocator: Allocator, policy: *const InboundPolicy, store: *const LabelStore) !PolicyEngine {
        var pe = try init(allocator, policy);
        pe.label_store = store;
        return pe;
    }

    /// Set or replace the label store.
    pub fn setLabelStore(self: *PolicyEngine, store: ?*const LabelStore) void {
        self.label_store = store;
    }

    pub fn deinit(self: *PolicyEngine) void {
        for (self.entries.items) |*e| {
            if (e.whitelist) |*w| {
                freeStringKeys(self.allocator, w);
                w.deinit(self.allocator);
            }
        }
        self.entries.deinit(self.allocator);
    }

    /// Check a peer's public key against the policy rules.
    pub fn check(self: *const PolicyEngine, pubkey: *const [32]u8) PolicyResult {
        var hex_buf: [64]u8 = undefined;
        const hex_chars = "0123456789abcdef";
        for (pubkey, 0..) |b, i| {
            hex_buf[i * 2] = hex_chars[b >> 4];
            hex_buf[i * 2 + 1] = hex_chars[b & 0x0f];
        }
        const pubkey_hex: []const u8 = &hex_buf;

        // Look up labels once (only if label store is set)
        const peer_labels: []const []const u8 = if (self.label_store) |store|
            store.getLabels(pubkey_hex)
        else
            &.{};

        for (self.entries.items) |*entry| {
            const rule = entry.rule;
            const match_type = rule.@"match".pubkey.@"type";

            var pubkey_matched = false;
            var needs_zgrlan = false;
            var zgrlan_peer: []const u8 = "";

            if (mem.eql(u8, match_type, "any")) {
                pubkey_matched = true;
            } else if (mem.eql(u8, match_type, "whitelist")) {
                if (entry.whitelist) |wl| {
                    if (wl.get(pubkey_hex) != null) {
                        pubkey_matched = true;
                    }
                }
            } else if (mem.eql(u8, match_type, "zgrlan")) {
                pubkey_matched = true;
                needs_zgrlan = true;
                zgrlan_peer = rule.@"match".pubkey.peer;
            } else {
                continue;
            }

            if (!pubkey_matched) continue;

            // If the rule has label patterns, the peer must also match at least one
            if (rule.@"match".labels.len > 0) {
                if (!matchLabels(peer_labels, rule.@"match".labels)) {
                    continue;
                }
            }

            return .{
                .action = rule.action,
                .services = rule.services,
                .rule_name = rule.name,
                .needs_zgrlan_verify = needs_zgrlan,
                .zgrlan_peer = zgrlan_peer,
            };
        }

        return .{
            .action = self.default_action,
            .services = &.{},
            .rule_name = "default",
            .needs_zgrlan_verify = false,
            .zgrlan_peer = "",
        };
    }

    pub fn reload(self: *PolicyEngine) !void {
        for (self.entries.items) |*entry| {
            if (entry.list_path.len > 0) {
                if (entry.whitelist) |*w| {
                    freeStringKeys(self.allocator, w);
                    w.deinit(self.allocator);
                }
                entry.whitelist = try loadPubkeyList(self.allocator, entry.list_path);
            }
        }
    }
};

/// Free all allocated key strings in a StringHashMap before deiniting it.
fn freeStringKeys(allocator: Allocator, map: *std.StringHashMapUnmanaged(void)) void {
    var it = map.iterator();
    while (it.next()) |entry| {
        allocator.free(entry.key_ptr.*);
    }
}

fn loadPubkeyList(allocator: Allocator, path: []const u8) !std.StringHashMapUnmanaged(void) {
    var keys = std.StringHashMapUnmanaged(void){};
    errdefer keys.deinit(allocator);

    const file = std.fs.cwd().openFile(path, .{}) catch |err| return err;
    defer file.close();
    const data = file.readToEndAlloc(allocator, 1024 * 1024) catch return error.IoError;
    defer allocator.free(data);

    var it = mem.splitScalar(u8, data, '\n');
    while (it.next()) |line| {
        const trimmed = mem.trim(u8, line, " \t\r");
        if (trimmed.len == 0 or trimmed[0] == '#') continue;
        if (trimmed.len != 64) return error.InvalidInput;
        for (trimmed) |c| {
            if (!((c >= '0' and c <= '9') or (c >= 'a' and c <= 'f') or (c >= 'A' and c <= 'F'))) {
                return error.InvalidInput;
            }
        }
        const lower = try allocator.alloc(u8, 64);
        for (trimmed, 0..) |c, i| {
            lower[i] = if (c >= 'A' and c <= 'Z') c + 32 else c;
        }
        try keys.put(allocator, lower, {});
    }
    return keys;
}

test "any match" {
    const rules = [_]InboundRule{.{
        .name = "allow-all",
        .@"match" = .{ .pubkey = .{ .@"type" = "any" } },
        .services = &.{.{ .proto = "*", .port = "*" }},
        .action = "allow",
    }};
    const policy = InboundPolicy{ .default = "deny", .rules = &rules };
    var pe = try PolicyEngine.init(std.testing.allocator, &policy);
    defer pe.deinit();

    const key = [_]u8{0xab} ** 32;
    const result = pe.check(&key);
    try std.testing.expectEqualStrings("allow", result.action);
    try std.testing.expectEqualStrings("allow-all", result.rule_name);
}

test "zgrlan match" {
    const rules = [_]InboundRule{.{
        .name = "company",
        .@"match" = .{ .pubkey = .{ .@"type" = "zgrlan", .peer = "company.zigor.net" } },
        .services = &.{.{ .proto = "tcp", .port = "80,443" }},
        .action = "allow",
    }};
    const policy = InboundPolicy{ .default = "deny", .rules = &rules };
    var pe = try PolicyEngine.init(std.testing.allocator, &policy);
    defer pe.deinit();

    const key = [_]u8{0xab} ** 32;
    const result = pe.check(&key);
    try std.testing.expectEqualStrings("allow", result.action);
    try std.testing.expect(result.needs_zgrlan_verify);
    try std.testing.expectEqualStrings("company.zigor.net", result.zgrlan_peer);
}

test "default deny" {
    const policy = InboundPolicy{ .default = "deny" };
    var pe = try PolicyEngine.init(std.testing.allocator, &policy);
    defer pe.deinit();

    const key = [_]u8{0} ** 32;
    const result = pe.check(&key);
    try std.testing.expectEqualStrings("deny", result.action);
    try std.testing.expectEqualStrings("default", result.rule_name);
}

test "default allow" {
    const policy = InboundPolicy{ .default = "allow" };
    var pe = try PolicyEngine.init(std.testing.allocator, &policy);
    defer pe.deinit();

    const key = [_]u8{0} ** 32;
    const result = pe.check(&key);
    try std.testing.expectEqualStrings("allow", result.action);
}

test "label match" {
    var store = LabelStore.init(std.testing.allocator);
    defer store.deinit();

    // key 0xab repeated = "abababab..." in hex
    const pk_hex = "abababababababababababababababababababababababababababababababab";
    try store.addLabels(pk_hex, &.{"host.zigor.net/trusted"});

    const rules = [_]InboundRule{.{
        .name = "label-trusted",
        .@"match" = .{
            .pubkey = .{ .@"type" = "any" },
            .labels = &.{"host.zigor.net/trusted"},
        },
        .services = &.{.{ .proto = "tcp", .port = "80,443" }},
        .action = "allow",
    }};
    const policy = InboundPolicy{ .default = "deny", .rules = &rules };
    var pe = try PolicyEngine.init(std.testing.allocator, &policy);
    defer pe.deinit();
    pe.setLabelStore(&store);

    const key = [_]u8{0xab} ** 32;
    const result = pe.check(&key);
    try std.testing.expectEqualStrings("allow", result.action);
    try std.testing.expectEqualStrings("label-trusted", result.rule_name);

    // Unknown key has no labels -> deny
    const unknown = [_]u8{0} ** 32;
    const result2 = pe.check(&unknown);
    try std.testing.expectEqualStrings("deny", result2.action);
}

test "no label store - label rules skip" {
    const rules = [_]InboundRule{.{
        .name = "label-only",
        .@"match" = .{
            .pubkey = .{ .@"type" = "any" },
            .labels = &.{"host.zigor.net/trusted"},
        },
        .services = &.{.{ .proto = "*", .port = "*" }},
        .action = "allow",
    }};
    const policy = InboundPolicy{ .default = "deny", .rules = &rules };
    var pe = try PolicyEngine.init(std.testing.allocator, &policy);
    defer pe.deinit();

    const key = [_]u8{0xab} ** 32;
    const result = pe.check(&key);
    try std.testing.expectEqualStrings("deny", result.action);
}
